### DISCUSSION TITLE
Maximum error between source geometry and resulting navmesh

### DISCUSSION THREAD

**Discussion by eugeneko** [recastnavigation/recastnavigation#689] [Q&A]
When I generate navmesh from my geometry, the difference between source geometry and navmesh varies from -2 to 2.5 units (1 unit = 1 meter for me), which is not acceptable. I use tile cache and obstacles, therefore detail mesh is not accessible for me.

How can I control amount of vertical error? I am ok to have more polygons if it means more precise navigation.
I can deal with error of [-0.5, 0.5] meters, I cannot deal with the error of almost [-2.5, 2.5] meters.

**Comment by grahamboree** [comment]
The navmesh geometry is ultimately generated from the voxelized representation of the geometry.  If the voxel size is large, the navmesh will be offset from the source geometry a noticeable amount.

It's hard to say more without additional information.  Could you post your rasterization settings?

**Reply by eugeneko** [reply]
I could post some settings, but is there some kind of heuristics formula?
Let's say I have `cs=0.15, ch=0.2`. What magnitude of vertical error can I expect?
`maxSimplificationError` is apparently irrelevant because it's about horizontal error, not vertical one.

**Reply by grahamboree** [reply]
I would expect it to be roughly within 0.15 of the source geometry, but there are a lot of factors here.  e.g. if the mesh simplification combines a ramp with a floor it can generate a plane that deviates much more than that.  It's important to keep in mind that this is okay for navmeshes, since their purpose is to provide a simplified world representation for pathfinding.  You shouldn't rely on navmeshes as a source of walkable surface geometry.

**Reply by eugeneko** [reply]
> You shouldn't rely on navmeshes as a source of walkable surface geometry.

But how should I tell where the path point is actually located? At the end of the day, I need to somehow move my objects along the path.

I currently solve it by raycasting to the surface, but this approach is not sustainable -- with [-2.5, 2.5] error I can accidentally sample upper floor with raycast, if I have two close walkable surfaces.

I don't need navmesh that matches walkable surface geometry, but I need to somehow build correlation between arbitrary navmesh point and corresponding surface geometry point.

**Reply by grahamboree** [reply]
Raycasting to the surface is an approach that should be workable.  It's really difficult to help much more without some additional context.  Screenshots, input geometry, rasterization settings, etc.

**Reply by eugeneko** [reply]
@grahamboree okay, let's make a scheme (side view).
![image](https://github.com/recastnavigation/recastnavigation/assets/8576192/bb74ac37-13cf-4650-ac9d-439ed0a6bd7f)

- I have horizontal heightmap-like surface of variable height
- Entiretry of this surface is perfectly navigable, there are no obstacles or non-walkable polygons, so it's basically a quad for navigation
- When navmesh is simplified, huge areas of this area may be reduced to the single polygon without care for error

It would be great if I could limit max error in such scenarios. Otherwise, error may be basically unlimited.
